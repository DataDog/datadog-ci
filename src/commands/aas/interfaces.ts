/**
 * Configuration options provided by the user through
 * the CLI in order to instrument properly.
 */
export interface AasConfigOptions {
  // AAS Targeting options
  subscriptionId?: string
  resourceGroup?: string
  aasName?: string
  resourceIds?: string[]

  // Configuration options
  service?: string
  environment?: string
  version?: string
  isInstanceLoggingEnabled?: boolean
  isProfilingEnabled?: boolean
  logPath?: string
  isDotnet?: boolean
  // no-dd-sa:typescript-best-practices/boolean-prop-naming
  shouldNotRestart?: boolean
}
