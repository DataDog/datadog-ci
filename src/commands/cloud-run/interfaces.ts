/**
 * Summarize relevant information about a Cloud Run log.
 * Also used to build CSV log files.
 * @typedef {Object} CloudRunLog
 * @property {string} severity - The level of severity of the log. It can be values such as 'DEBUG', 'INFO', 'ERROR', etc.
 * @property {string} timestamp - The timestamp of when the log was generated.
 * @property {string} logName - The name of the log
 * @property {string} message - The actual log message detailing what event occurred.
 */
export interface CloudRunLog {
  severity: string
  timestamp: string
  logName: string
  message: string
}

/**
 * Contains all the information used to create a log file.
 * @typedef {Object} LogConfig
 * @property {string} type - string name of the type of log. Used when printing CLI messages.
 * @property {string} fileName - The name of the log file (such as 'all_logs.csv' or 'error_logs.csv')
 * @property {string} [severityFilter] - Optional filter to modify the Logging query. Example: ' AND severity="DEBUG"'
 */
export interface LogConfig {
  type: string
  fileName: string
  severityFilter?: string
}
