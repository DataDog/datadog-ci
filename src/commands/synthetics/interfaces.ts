import {Metadata} from '../../helpers/interfaces'
import {ProxyConfiguration} from '../../helpers/utils'

interface Timings {
  dns: number
  download: number
  firstByte: number
  ssl: number
  tcp: number
  total: number
}

export interface Result {
  device: {
    id: string
  }
  duration?: number
  error?: string
  errorCode?: string
  errorMessage?: string
  eventType: string
  passed: boolean
  stepDetails: Step[]
  timings?: Timings
  unhealthy?: boolean
}

export interface PollResult {
  dc_id: number
  result: Result
  resultID: string
}

interface Resource {
  duration: number
  size: number
  type: string
  url: string
}

export interface Step {
  apmTraceIds: string[]
  browserErrors: string[]
  description: string
  duration: number
  error?: string
  resource: Resource
  screenshotBucketKey: boolean
  skipped: boolean
  snapshotBucketKey: boolean
  stepId: number
  type: string
  url: string
  value: string
}

export interface Test {
  config: {
    assertions: Assertion[]
    request: {
      headers: {[key: string]: string}
      method: string
      timeout: number
      url: string
    }
    variables: string[]
  }
  created_at: string
  created_by: User
  locations: string[]
  message: string
  modified_at: string
  modified_by: User
  monitor_id: number
  name: string
  options: {
    ci?: {
      executionRule: ExecutionRule
    }
    device_ids: string[]
    min_failure_duration: number
    min_location_failed: number
    tick_every: number
  }
  overall_state: number
  overall_state_modified: string
  public_id: string
  status: string
  stepCount: number
  tags: string[]
  type: string
}

export interface Assertion {
  actual: string | number | Date | {[key: string]: any}
  errorMessage?: string
  operator: Operator
  property?: string
  target: string | number | Date | {[key: string]: any}
  type: string
  valid: boolean
}

export enum Operator {
  contains = 'contains',
  doesNotContain = 'doesNotContain',
  is = 'is',
  isNot = 'isNot',
  lessThan = 'lessThan',
  matches = 'matches',
  doesNotMatch = 'doesNotMatch',
  validates = 'validates',
  isInMoreThan = 'isInMoreThan',
  isInLessThan = 'isInLessThan',
}

interface User {
  email: string
  handle: string
  id: number
  name: string
}

export interface TriggerResponse {
  device: string
  location: number
  public_id: string
  result_id: string
}

export interface TriggerResult extends TriggerResponse {
  pollingTimeout: number
  result?: PollResult
}

interface Location {
  display_name: string
  id: number
  is_active: boolean
  name: string
  region: string
}

export interface LocationsMapping {
  [key: number]: string
}

export interface Trigger {
  locations: Location[]
  results: TriggerResponse[]
  triggered_check_ids: string[]
}

interface RetryConfig {
  count: number
  interval: number
}

export interface ConfigOverride {
  allowInsecureCertificates?: boolean
  basicAuth?: BasicAuthCredentials
  body?: string
  bodyType?: string
  cookies?: string
  deviceIds?: string[]
  executionRule?: ExecutionRule
  followRedirects?: boolean
  headers?: {[key: string]: string}
  locations?: string[]
  pollingTimeout?: number
  retry?: RetryConfig
  startUrl?: string
  variables?: {[key: string]: string}
}

export interface Payload extends ConfigOverride {
  metadata?: Metadata
  public_id: string
}

interface BasicAuthCredentials {
  password: string
  username: string
}

export interface TemplateContext extends NodeJS.ProcessEnv {
  DOMAIN: string
  HOST: string
  HOSTNAME: string
  ORIGIN: string
  PARAMS: string
  PATHNAME: string
  PORT: string
  PROTOCOL: string
  SUBDOMAIN: string | undefined
  URL: string
}

export interface TriggerConfig {
  config: ConfigOverride
  id: string
}

export enum ExecutionRule {
  BLOCKING = 'blocking',
  NON_BLOCKING = 'non_blocking',
  SKIPPED = 'skipped',
}

export interface Suite {
  tests: TriggerConfig[]
}

export interface TestSearchResult {
  tests: {
    public_id: string
  }[]
}

export interface APIHelper {
  getTest(testId: string): Promise<Test>
  pollResults(resultIds: string[]): Promise<{results: PollResult[]}>
  searchTests(query: string): Promise<TestSearchResult>
  triggerTests(testsToTrigger: Payload[]): Promise<Trigger>
}

export interface APIConfiguration {
  apiKey: string
  appKey: string
  baseIntakeUrl: string
  baseUrl: string
  proxyOpts: ProxyConfiguration
}
