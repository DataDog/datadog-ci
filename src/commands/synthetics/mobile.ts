import * as crypto from 'crypto'
import fs from 'fs'

import {APIHelper, EndpointError, formatBackendErrors, getApiHelper} from './api'
import {CiError} from './errors'
import {
  MobileApplicationUploadPart,
  MobileApplicationUploadPartResponse,
  MobileApplicationVersion,
  MultipartPresignedUrlsResponse,
  Test,
  TestPayload,
  UploadApplicationCommandConfig,
  UserConfigOverride,
} from './interfaces'

const UPLOAD_FILE_MAX_PART_SIZE = 10 * 1024 * 1024 // MiB

export const getSizeAndPartsFromFile = async (
  filePath: string
): Promise<{appSize: number; parts: MobileApplicationUploadPart[]}> => {
  const readStreamOptions = {
    // Limit the chunk size for the stream
    // https://nodejs.org/api/stream.html#buffering
    highWaterMark: UPLOAD_FILE_MAX_PART_SIZE,
  }
  const fileStream = fs.createReadStream(filePath, readStreamOptions)
  const parts: MobileApplicationUploadPart[] = []
  for await (const chunk of fileStream) {
    if (!(chunk instanceof Buffer)) {
      // this should never happen, but for-await-of creates an any that we don't want
      throw new Error('Unexpected chunk type from file stream')
    }

    parts.push({
      md5: crypto.createHash('md5').update(chunk).digest('base64'),
      partNumber: parts.length + 1,
      blob: chunk,
    })
  }

  return {
    appSize: fileStream.bytesRead,
    parts,
  }
}

export const uploadMobileApplications = async (
  api: APIHelper,
  applicationPathToUpload: string,
  mobileApplicationId: string
): Promise<string> => {
  const {appSize, parts} = await getSizeAndPartsFromFile(applicationPathToUpload)

  let multipartPresignedUrlsResponse: MultipartPresignedUrlsResponse
  try {
    multipartPresignedUrlsResponse = await api.getMobileApplicationPresignedURLs(mobileApplicationId, appSize, parts)
  } catch (e) {
    throw new EndpointError(`Failed to get presigned URL: ${formatBackendErrors(e)}\n`, e.response?.status)
  }

  let uploadPartResponses: MobileApplicationUploadPartResponse[]
  try {
    uploadPartResponses = await api.uploadMobileApplicationPart(
      parts,
      multipartPresignedUrlsResponse.multipart_presigned_urls_params
    )
  } catch (e) {
    throw new EndpointError(`Failed to upload mobile application: ${formatBackendErrors(e)}\n`, e.response?.status)
  }

  const {upload_id: uploadId, key} = multipartPresignedUrlsResponse.multipart_presigned_urls_params
  try {
    await api.completeMultipartMobileApplicationUpload(mobileApplicationId, uploadId, key, uploadPartResponses)
  } catch (e) {
    throw new EndpointError(
      `Failed to complete upload mobile application: ${formatBackendErrors(e)}\n`,
      e.response?.status
    )
  }

  return multipartPresignedUrlsResponse.file_name
}

export const uploadApplication = async (
  api: APIHelper,
  applicationPathToUpload: string,
  testApplicationId: string,
  uploadedApplicationByPath: {[applicationFilePath: string]: {applicationId: string; fileName: string}[]}
) => {
  const fileName = await uploadMobileApplications(api, applicationPathToUpload, testApplicationId)
  if (!(applicationPathToUpload in uploadedApplicationByPath)) {
    uploadedApplicationByPath[applicationPathToUpload] = []
  }

  uploadedApplicationByPath[applicationPathToUpload].push({
    applicationId: testApplicationId,
    fileName,
  })
}

export const overrideMobileConfig = (
  userConfigOverride: UserConfigOverride,
  overriddenTest: TestPayload,
  test: Test,
  localApplicationOverride?: {applicationId: string; fileName: string}
) => {
  if (localApplicationOverride) {
    overriddenTest.mobileApplication = {
      applicationId: localApplicationOverride.applicationId,
      referenceId: localApplicationOverride.fileName,
      referenceType: 'temporary',
    }
  } else if (userConfigOverride.mobileApplicationVersion) {
    overriddenTest.mobileApplication = {
      applicationId: test.options.mobileApplication!.applicationId,
      referenceId: userConfigOverride.mobileApplicationVersion,
      referenceType: 'version',
    }
  }
}

export const shouldUploadApplication = (
  applicationPathToUpload: string,
  testApplicationId: string,
  uploadedApplicationByPath: {[applicationFilePath: string]: {applicationId: string; fileName: string}[]}
): boolean =>
  !(applicationPathToUpload in uploadedApplicationByPath) ||
  !uploadedApplicationByPath[applicationPathToUpload].some(({applicationId}) => applicationId === testApplicationId)

export const uploadApplicationAndOverrideConfig = async (
  api: APIHelper,
  test: Test,
  userConfigOverride: UserConfigOverride,
  overriddenTestsToTrigger: TestPayload,
  uploadedApplicationByPath: {[applicationFilePath: string]: {applicationId: string; fileName: string}[]}
): Promise<void> => {
  const testApplicationId = test.options.mobileApplication!.applicationId
  if (
    userConfigOverride.mobileApplicationVersionFilePath &&
    shouldUploadApplication(
      userConfigOverride.mobileApplicationVersionFilePath,
      testApplicationId,
      uploadedApplicationByPath
    )
  ) {
    await uploadApplication(
      api,
      userConfigOverride.mobileApplicationVersionFilePath,
      testApplicationId,
      uploadedApplicationByPath
    )
  }

  const localApplicationOverride = userConfigOverride.mobileApplicationVersionFilePath
    ? uploadedApplicationByPath[userConfigOverride.mobileApplicationVersionFilePath].find(
        ({applicationId}) => applicationId === testApplicationId
      )
    : undefined

  overrideMobileConfig(userConfigOverride, overriddenTestsToTrigger, test, localApplicationOverride)
}

export const createNewMobileVersion = async (
  api: APIHelper,
  version: MobileApplicationVersion
): Promise<MobileApplicationVersion> => {
  let newVersion: MobileApplicationVersion
  try {
    newVersion = await api.createMobileVersion(version)
  } catch (e) {
    throw new EndpointError(`Failed create new Mobile Version: ${formatBackendErrors(e)}\n`, e.response?.status)
  }

  return newVersion
}

export const uploadMobileApplicationVersion = async (
  config: UploadApplicationCommandConfig
): Promise<MobileApplicationVersion> => {
  const api = getApiHelper(config)

  if (!config.mobileApplicationVersionFilePath) {
    throw new CiError('MISSING_MOBILE_APPLICATION_PATH', 'Mobile application path is required.')
  }

  if (!config.mobileApplicationId) {
    throw new CiError('MISSING_MOBILE_APPLICATION_ID', 'Mobile application id is required.')
  }

  if (!config.versionName) {
    throw new CiError('MISSING_MOBILE_VERSION_NAME', 'Version name is required')
  }
  config.latest = config.latest ?? false

  const fileName = await uploadMobileApplications(
    api,
    config.mobileApplicationVersionFilePath,
    config.mobileApplicationId
  )

  const version = await createNewMobileVersion(api, {
    file_name: fileName,
    application_id: config.mobileApplicationId,
    original_file_name: config.mobileApplicationVersionFilePath,
    version_name: config.versionName,
    is_latest: config.latest,
  })

  return version
}
