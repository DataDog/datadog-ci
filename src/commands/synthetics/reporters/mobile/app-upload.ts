import chalk from 'chalk'
import ora from 'ora'

import {CommandContext} from '../../../../helpers/interfaces'

import {AppUploadDetails} from '../../interfaces'

import {ICONS} from '../constants'

export class AppUploadReporter {
  private context: CommandContext
  private spinner?: ora.Ora
  private startTime: number

  constructor(context: CommandContext) {
    this.context = context
    this.startTime = Date.now()
  }

  public start(appsToUpload: AppUploadDetails[], prependLineBreak = false): void {
    this.write(`${prependLineBreak ? '\n' : ''}${appsToUpload.length} mobile application(s) to upload:\n`)
    this.write(appsToUpload.map((appToUpload) => this.getAppDescription(appToUpload)).join('\n') + '\n')
  }

  public renderProgress(numberOfApplicationsLeft: number): void {
    const text = `Uploading ${numberOfApplicationsLeft} application(s)â€¦`
    this.spinner?.stop()
    this.spinner = ora({
      stream: this.context.stdout,
      text,
    })
    this.spinner.start()
  }

  public reportSuccess(): void {
    this.endRendering()
    this.write(
      `\n${ICONS.SUCCESS} Successfully uploaded in ${Math.round((Date.now() - this.startTime) / 1000)} seconds\n`
    )
  }

  public reportFailure(failedApp: AppUploadDetails): void {
    this.endRendering()
    this.write(`\n${ICONS.FAILED} Failed to upload application:\n${this.getAppDescription(failedApp)}\n`)
  }

  public endRendering(): void {
    this.spinner?.stop()
    delete this.spinner
  }

  private getAppDescription(appUploadDetails: AppUploadDetails): string {
    let versionPrepend = ''
    if (appUploadDetails.versionName) {
      versionPrepend = `Version ${chalk.dim.cyan(appUploadDetails.versionName)} - `
    }

    return `    ${versionPrepend}Application ID ${chalk.dim.cyan(appUploadDetails.appId)} - Local Path ${chalk.dim.cyan(
      appUploadDetails.appPath
    )}`
  }

  private write(message: string): void {
    this.context.stdout.write(message)
  }
}
