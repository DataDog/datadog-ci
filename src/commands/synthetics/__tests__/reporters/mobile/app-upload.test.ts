jest.unmock('chalk')

// Force `process.platform` to have the same spinner snapshots on all platforms.
// Without this, `ora` falls back to use a `-` on Windows, resulting in snapshot diffs (`-` instead of `â ‹`).
// See https://github.com/sindresorhus/ora#spinner
Object.defineProperty(process, 'platform', {
  value: 'linux',
  configurable: true,
  writable: true,
})

import {CommandContext} from '../../../../../helpers/interfaces'

import {AppUploadReporter} from '../../../reporters/mobile/app-upload'

/* eslint-disable jest/no-conditional-expect */
describe('AppUploadReporter', () => {
  let initialCiEnv: string | undefined
  let simulatedTerminalOutput: string
  let writeMock: jest.Mock
  let clearLine: jest.Mock
  let ttyReporter: AppUploadReporter

  beforeEach(() => {
    jest.useFakeTimers()
    initialCiEnv = process.env.CI

    simulatedTerminalOutput = ''

    writeMock = jest.fn().mockImplementation((text: string) => {
      // Ignore show/hide cursor ANSI codes.
      if (text.match(/\u001b\[\?25(l|h)/)) {
        return
      }

      simulatedTerminalOutput += text
    })

    clearLine = jest.fn().mockImplementation(() => {
      const allLinesMinusLast = simulatedTerminalOutput.split('\n').slice(0, -1)
      allLinesMinusLast.push('')
      simulatedTerminalOutput = allLinesMinusLast.join('\n')
    })

    const ttyContext: unknown = {
      stdout: {
        isTTY: true,
        write: writeMock,
        clearLine,
        cursorTo: jest.fn(),
        moveCursor: jest.fn(),
      },
      stderr: {
        write: writeMock,
      },
    }

    ttyReporter = new AppUploadReporter(ttyContext as CommandContext)
  })

  afterEach(() => {
    jest.useRealTimers()
    if (initialCiEnv !== undefined) {
      process.env.CI = initialCiEnv
    } else {
      delete process.env.CI
    }
  })

  describe('start', () => {
    test('outputs apps to upload', async () => {
      const appsToUpload = [
        {
          versionName: '1.0.0',
          appId: '123',
          appPath: '/path/to/app',
        },
        {
          versionName: '2.0.0',
          appId: '456',
          appPath: '/path/to/another/app',
        },
      ]
      ttyReporter.start(appsToUpload)
      const output = writeMock.mock.calls.map((c) => c[0]).join('\n')
      expect(output).toMatchSnapshot()

      writeMock.mockClear()
      ttyReporter.start(appsToUpload, true)
      const output2 = writeMock.mock.calls.map((c) => c[0]).join('\n')
      expect(output2).toMatchSnapshot()
    })
  })

  describe('renderProgress', () => {
    test.each([false, true])('Outputs progress (in CI: %s)', async (inCI) => {
      if (inCI) {
        process.env.CI = 'true'
      } else {
        delete process.env.CI
      }

      ttyReporter.renderProgress(2)
      expect(clearLine).not.toHaveBeenCalled()
      expect(simulatedTerminalOutput).toMatchSnapshot()

      ttyReporter.renderProgress(1)
      if (inCI) {
        // In CI, there is no spinning, so `stopping` the spinner (which happens when rendering a 2nd time) does not clear a line.
        expect(clearLine).not.toHaveBeenCalled()
      } else {
        expect(clearLine).toHaveBeenCalled()
      }
      expect(simulatedTerminalOutput).toMatchSnapshot()
    })
  })

  describe('reportSuccess', () => {
    test.each([false, true])('Outputs success message (in CI: %s)', async (inCI) => {
      if (inCI) {
        process.env.CI = 'true'
      } else {
        delete process.env.CI
      }
      ttyReporter.renderProgress(1)
      ttyReporter.reportSuccess()
      expect(simulatedTerminalOutput).toMatchSnapshot()
      if (inCI) {
        // In CI, there is no spinning, so `stopping` the spinner (which happens when rendering a 2nd time) does not clear a line.
        expect(clearLine).not.toHaveBeenCalled()
      } else {
        expect(clearLine).toHaveBeenCalled()
      }
    })
  })

  describe('reportFailure', () => {
    test.each([false, true])('Outputs failure message (in CI: %s)', async (inCI) => {
      if (inCI) {
        process.env.CI = 'true'
      } else {
        delete process.env.CI
      }
      ttyReporter.renderProgress(1)
      ttyReporter.reportFailure({
        versionName: '1.0.0',
        appId: '123',
        appPath: '/path/to/app',
      })
      expect(simulatedTerminalOutput).toMatchSnapshot()
      if (inCI) {
        // In CI, there is no spinning, so `stopping` the spinner (which happens when rendering a 2nd time) does not clear a line.
        expect(clearLine).not.toHaveBeenCalled()
      } else {
        expect(clearLine).toHaveBeenCalled()
      }
    })
  })

  describe('endRendering', () => {
    test.each([false, true])('Ends rendering (in CI: %s)', async (inCI) => {
      if (inCI) {
        process.env.CI = 'true'
      } else {
        delete process.env.CI
      }
      ttyReporter.renderProgress(1)
      expect(ttyReporter['spinner']).toBeDefined()
      ttyReporter.endRendering()
      expect(ttyReporter['spinner']).toBeUndefined()
      expect(simulatedTerminalOutput).toMatchSnapshot()
      if (inCI) {
        // In CI, there is no spinning, so `stopping` the spinner (which happens when rendering a 2nd time) does not clear a line.
        expect(clearLine).not.toHaveBeenCalled()
      } else {
        expect(clearLine).toHaveBeenCalled()
      }
    })
  })

  describe('integrationTests', () => {
    test.each([false, true])('Outputs a series of successful uploads correctly (in CI: %s)', async (inCI) => {
      if (inCI) {
        process.env.CI = 'true'
      } else {
        delete process.env.CI
      }
      ttyReporter.start([
        {
          versionName: '1.0.0',
          appId: '123',
          appPath: '/path/to/app',
        },
        {
          versionName: '2.0.0',
          appId: '456',
          appPath: '/path/to/another/app',
        },
      ])
      expect(simulatedTerminalOutput).toMatchSnapshot()

      ttyReporter.renderProgress(2)
      expect(simulatedTerminalOutput).toMatchSnapshot()

      ttyReporter.renderProgress(1)
      expect(simulatedTerminalOutput).toMatchSnapshot()

      ttyReporter.reportSuccess()
      expect(simulatedTerminalOutput).toMatchSnapshot()
    })
  })

  test.each([false, true])('Outputs a failure during uploads correctly (in CI: %s)', async (inCI) => {
    if (inCI) {
      process.env.CI = 'true'
    } else {
      delete process.env.CI
    }
    ttyReporter.start([
      {
        versionName: '1.0.0',
        appId: '123',
        appPath: '/path/to/app',
      },
      {
        versionName: '2.0.0',
        appId: '456',
        appPath: '/path/to/another/app',
      },
    ])
    expect(simulatedTerminalOutput).toMatchSnapshot()

    ttyReporter.renderProgress(2)
    expect(simulatedTerminalOutput).toMatchSnapshot()

    ttyReporter.renderProgress(1)
    expect(simulatedTerminalOutput).toMatchSnapshot()

    ttyReporter.reportFailure({
      versionName: '2.0.0',
      appId: '456',
      appPath: '/path/to/another/app',
    })
    expect(simulatedTerminalOutput).toMatchSnapshot()
  })
})
