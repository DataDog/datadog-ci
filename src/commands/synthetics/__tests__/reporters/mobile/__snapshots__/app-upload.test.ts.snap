// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppUploadReporter Outputs a failure during uploads correctly (in CI: false) 1`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
"
`;

exports[`AppUploadReporter Outputs a failure during uploads correctly (in CI: false) 2`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
[36mâ ‹[39m Uploading 2 application(s)â€¦"
`;

exports[`AppUploadReporter Outputs a failure during uploads correctly (in CI: false) 3`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
[36mâ ‹[39m Uploading 1 application(s)â€¦"
`;

exports[`AppUploadReporter Outputs a failure during uploads correctly (in CI: false) 4`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m

[1m[31mâœ–[39m[22m Failed to upload application:
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
"
`;

exports[`AppUploadReporter Outputs a failure during uploads correctly (in CI: true) 1`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
"
`;

exports[`AppUploadReporter Outputs a failure during uploads correctly (in CI: true) 2`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
- Uploading 2 application(s)â€¦
"
`;

exports[`AppUploadReporter Outputs a failure during uploads correctly (in CI: true) 3`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
- Uploading 2 application(s)â€¦
- Uploading 1 application(s)â€¦
"
`;

exports[`AppUploadReporter Outputs a failure during uploads correctly (in CI: true) 4`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
- Uploading 2 application(s)â€¦
- Uploading 1 application(s)â€¦

[1m[31mâœ–[39m[22m Failed to upload application:
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
"
`;

exports[`AppUploadReporter endRendering Ends rendering (in CI: false) 1`] = `""`;

exports[`AppUploadReporter endRendering Ends rendering (in CI: true) 1`] = `
"- Uploading 1 application(s)â€¦
"
`;

exports[`AppUploadReporter integrationTests Outputs a series of successful uploads correctly (in CI: false) 1`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
"
`;

exports[`AppUploadReporter integrationTests Outputs a series of successful uploads correctly (in CI: false) 2`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
[36mâ ‹[39m Uploading 2 application(s)â€¦"
`;

exports[`AppUploadReporter integrationTests Outputs a series of successful uploads correctly (in CI: false) 3`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
[36mâ ‹[39m Uploading 1 application(s)â€¦"
`;

exports[`AppUploadReporter integrationTests Outputs a series of successful uploads correctly (in CI: false) 4`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m

[1m[32mâœ“[39m[22m Successfully uploaded in 0 seconds
"
`;

exports[`AppUploadReporter integrationTests Outputs a series of successful uploads correctly (in CI: true) 1`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
"
`;

exports[`AppUploadReporter integrationTests Outputs a series of successful uploads correctly (in CI: true) 2`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
- Uploading 2 application(s)â€¦
"
`;

exports[`AppUploadReporter integrationTests Outputs a series of successful uploads correctly (in CI: true) 3`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
- Uploading 2 application(s)â€¦
- Uploading 1 application(s)â€¦
"
`;

exports[`AppUploadReporter integrationTests Outputs a series of successful uploads correctly (in CI: true) 4`] = `
"2 mobile application(s) to upload:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
- Uploading 2 application(s)â€¦
- Uploading 1 application(s)â€¦

[1m[32mâœ“[39m[22m Successfully uploaded in 0 seconds
"
`;

exports[`AppUploadReporter renderProgress Outputs progress (in CI: false) 1`] = `"[36mâ ‹[39m Uploading 2 application(s)â€¦"`;

exports[`AppUploadReporter renderProgress Outputs progress (in CI: false) 2`] = `"[36mâ ‹[39m Uploading 1 application(s)â€¦"`;

exports[`AppUploadReporter renderProgress Outputs progress (in CI: true) 1`] = `
"- Uploading 2 application(s)â€¦
"
`;

exports[`AppUploadReporter renderProgress Outputs progress (in CI: true) 2`] = `
"- Uploading 2 application(s)â€¦
- Uploading 1 application(s)â€¦
"
`;

exports[`AppUploadReporter reportFailure Outputs failure message (in CI: false) 1`] = `
"
[1m[31mâœ–[39m[22m Failed to upload application:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
"
`;

exports[`AppUploadReporter reportFailure Outputs failure message (in CI: true) 1`] = `
"- Uploading 1 application(s)â€¦

[1m[31mâœ–[39m[22m Failed to upload application:
    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
"
`;

exports[`AppUploadReporter reportSuccess Outputs success message (in CI: false) 1`] = `
"
[1m[32mâœ“[39m[22m Successfully uploaded in 0 seconds
"
`;

exports[`AppUploadReporter reportSuccess Outputs success message (in CI: true) 1`] = `
"- Uploading 1 application(s)â€¦

[1m[32mâœ“[39m[22m Successfully uploaded in 0 seconds
"
`;

exports[`AppUploadReporter start outputs apps to upload 1`] = `
"2 mobile application(s) to upload:

    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
"
`;

exports[`AppUploadReporter start outputs apps to upload 2`] = `
"
2 mobile application(s) to upload:

    Version [2m[36m1.0.0[39m[22m - Application ID [2m[36m123[39m[22m - Local Path [2m[36m/path/to/app[39m[22m
    Version [2m[36m2.0.0[39m[22m - Application ID [2m[36m456[39m[22m - Local Path [2m[36m/path/to/another/app[39m[22m
"
`;
